library/ceph_fs.py

```diff
@@ -0,0 +1,339 @@
+    '''
+    Set parameter to a fs
+    '''
+
+    cluster = module.params.get('cluster')
+    name = module.params.get('name')
+    max_mds = module.params.get('max_mds')
+
+    args = ['set', name, 'max_mds', str(max_mds)]
+
+    cmd = generate_ceph_cmd(cluster=cluster, args=args, container_image=container_image)
+
+    return cmd
+
+
+def exit_module(module, out, rc, cmd, err, startd, changed=False):
+    endd = datetime.datetime.now()
+    delta = endd - startd
+
+    result = dict(
+        cmd=cmd,
+        start=str(startd),
+        end=str(endd),
+        delta=str(delta),
+        rc=rc,
+        stdout=out.rstrip("\r\n"),
+        stderr=err.rstrip("\r\n"),
+        changed=changed,
+    )
+    module.exit_json(**result)
+
+
+def run_module():
+    module_args = dict(
+        cluster=dict(type='str', required=False, default='ceph'),
+        name=dict(type='str', required=True),
+        state=dict(type='str', required=False, choices=['present', 'absent', 'info'], default='present'),
+        data=dict(type='str', required=False),
+        metadata=dict(type='str', required=False),
+        max_mds=dict(type='int', required=False),
+    )
+
+    module = AnsibleModule(
+        argument_spec=module_args,
+        supports_check_mode=True,
+        required_if=[['state', 'present', ['data', 'metadata']]],
+    )
+
+    # Gather module parameters in variables
+    name = module.params.get('name')
+    state = module.params.get('state')
+    max_mds = module.params.get('max_mds')
+
+    if module.check_mode:
+        module.exit_json(
+            changed=False,
+            stdout='',
+            stderr='',
+            rc=0,
+            start='',
+            end='',
+            delta='',
+        )
+
+    startd = datetime.datetime.now()
+    changed = False
+
+    # will return either the image name or None
+    container_image = is_containerized()
+
+    if state == "present":
+        rc, cmd, out, err = exec_commands(module, get_fs(module, container_image=container_image))
+        if rc == 0:
+            fs = json.loads(out)
+            if max_mds and fs["mdsmap"]["max_mds"] != max_mds:
+                rc, cmd, out, err = exec_commands(module, set_fs(module, container_image=container_image))
+                changed = True
+        else:
+            rc, cmd, out, err = exec_commands(module, create_fs(module, container_image=container_image))
+            if max_mds and max_mds > 1:
+                exec_commands(module, set_fs(module, container_image=container_image))
+            changed = True
+
+    elif state == "absent":
+        rc, cmd, out, err = exec_commands(module, get_fs(module, container_image=container_image))
+        if rc == 0:
+            rc, cmd, out, err = exec_commands(module, fail_fs(module, container_image=container_image))
+            rc, cmd, out, err = exec_commands(module, remove_fs(module, container_image=container_image))
+            changed = True
+        else:
+            rc = 0
```

Comment:
Although it won't hurt, I'm unsure enforcing `rc = 0` is really needed here: it is expected `state: absent` on a non-existing cephfs produces an rc != 0, that doesn't mean the module is failing, what is important is that `changed` is `False`

Reply:
no we need rc = 0 because the `ceph fs get xxx` doesn't return rc 0 when the filesystem doesn't exist so the module fails without this